org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ProjectUtils.java

```diff
@@ -385,6 +396,36 @@ private static boolean isBinary(Path file) {
 				&& !fileName.endsWith(SOURCE_JAR_SUFFIX));
 	}
 
+	public static IPath resolveGlobPath(IPath base, String glob) {
+		IPath pattern = new org.eclipse.core.runtime.Path(glob);
+		IPath baseDir = pattern.isAbsolute() ? org.eclipse.core.runtime.Path.ROOT : base;
+		return baseDir.append(pattern); // Append cwd to relative path
+	}
+
+	private static Map<Path, Set<String>> groupGlobsByPrefix(IPath base, Set<String> globs) {
```

Comment:
When using a list of globs to search files, they are organized in this way:
1. All globs are resolved to absolute paths by `resolveGlobPath`. In following work `${workspaceFolder:name}` will also be resolved in this method.
2. Globs are grouped in a <prefix, remain> map. Prefix is the longest segements that don't contain a special char. e.g., `/home/user/proj/lib/**/*.jar` will be split into `/home/user/proj/lib` and `**/*.jar`.
 
Then,
1. If the prefix is an absolute path to a regular file, then just stat its information.
2. Otherwise, search the path down using `DirectoryScanner`, with prefix as base directory. 

These are some basic optimizations inspired from `node-glob`.